cmd/rootMain.go

```diff
@@ -195,11 +86,117 @@ func rootMain(cmd *cobra.Command, args []string) exitCode {
 Try 'help'`)
 	}
 
-	exit := runShell(cmd.Commands(), rundir, mountpath, socketpath, execfile)
+	if !symlinkWash(rundir) {
+		return exitCode{1}
+	}
+
+	subc := listSubcommands(cmd.Commands())
+	comm, err := shell.GetShell().Command(subc, rundir)
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+
+	if execfile != "" {
+		file, err := os.Open(execfile)
+		if err != nil {
+			cmdutil.ErrPrintf("Error reading file %v: %v\n", execfile, err)
+			return exitCode{1}
+		}
+		comm.Stdin = file
+		defer file.Close()
+	} else if rootCommandFlag != "" {
+		comm.Stdin = strings.NewReader(rootCommandFlag)
+	} else {
+		comm.Stdin = os.Stdin
+	}
+	comm.Stdout = os.Stdout
+	comm.Stderr = os.Stderr
+	if comm.Env == nil {
+		comm.Env = os.Environ()
+	}
+	comm.Env = append(comm.Env,
+		"WASH_SOCKET="+socketpath,
+		"PATH="+rundir+string(os.PathListSeparator)+os.Getenv("PATH"),
+	)
+	comm.Dir = mountpath
+
+	var exit exitCode
+	if runErr := comm.Run(); runErr != nil {
+		if exitErr, ok := runErr.(*exec.ExitError); ok {
+			exit.value = exitErr.ExitCode()
+		}
+		cmdutil.ErrPrintf("%v\n", runErr)
+		exit.value = 1
+	}
 
 	srv.Stop()
 	if plugin.IsInteractive() {
 		cmdutil.Println("Goodbye!")
 	}
 	return exit
 }
+
+func makeCacheDir() (cachedir string, ok bool) {
+	var err error
+	if cachedir, err = os.UserCacheDir(); err != nil {
+		cmdutil.ErrPrintf("Unable to get user cache dir: %v\n", err)
+		return
+	}
+	cachedir = filepath.Join(cachedir, "wash")
+
+	// ensure cache directory exists
+	if err = os.MkdirAll(cachedir, 0750); err != nil {
+		cmdutil.ErrPrintf("Unable to create cache dir %v: %v\n", cachedir, err)
+		return
+	}
+	ok = true
+	return
+}
+
+func listSubcommands(subcommands []*cobra.Command) []string {
```

Comment:
Nit: `flattenSubcommands` ? Bit weird seeing the things that are being listed as a parameter to the function listing them.